driver/src/orderbook/streamed/balance.rs

```diff
@@ -63,29 +89,91 @@ impl Balance {
             // Saturating because withdraw requests can be for amounts larger than balance.
             balance = balance.saturating_sub(self.withdraw.amount);
         }
         Ok(balance)
     }
 
-    fn balance_with_deposit(&self, current_batch_id: BatchId) -> Result<U256> {
+    pub fn sell(&mut self, amount: u128, batch_id: BatchId) -> Result<()> {
+        self.sell_buy_internal(amount, batch_id, TradeType::Sell)
+    }
+
+    pub fn buy(&mut self, amount: u128, batch_id: BatchId) -> Result<()> {
+        self.sell_buy_internal(amount, batch_id, TradeType::Buy)
+    }
+
+    pub fn revert_sell(&mut self, amount: u128, batch_id: BatchId) -> Result<()> {
+        self.revert_sell_buy_internal(amount, batch_id, TradeType::Sell)
+    }
+
+    pub fn revert_buy(&mut self, amount: u128, batch_id: BatchId) -> Result<()> {
+        self.revert_sell_buy_internal(amount, batch_id, TradeType::Buy)
+    }
+
+    fn sell_buy_internal(
+        &mut self,
+        amount: u128,
+        batch_id: BatchId,
+        operation: TradeType,
+    ) -> Result<()> {
+        ensure!(self.proceeds.batch_id <= batch_id, "trade for past batch");
+        self.apply_existing_deposit_and_proceeds(batch_id)?;
+        // Now old proceeds have been cleared. If there is an existing proceed for this batch then
+        // setting the batch_id does nothing and we add to the field.
+        self.proceeds.batch_id = batch_id;
+        let field = self.proceeds.get_field(operation);
+        *field = field
+            .checked_add(amount.into())
+            .ok_or_else(|| anyhow!("math overflow"))?;
+        Ok(())
+    }
+
+    fn revert_sell_buy_internal(
+        &mut self,
+        amount: u128,
+        batch_id: BatchId,
+        operation: TradeType,
+    ) -> Result<()> {
+        ensure!(
+            self.proceeds.batch_id == batch_id,
+            "reverting non existent trade"
+        );
+        let field = self.proceeds.get_field(operation);
+        *field = field
+            .checked_sub(amount.into())
+            .ok_or_else(|| anyhow!("math underflow"))?;
+        Ok(())
+    }
+
+    fn balance_with_deposit_and_proceeds(&self, current_batch_id: BatchId) -> Result<U256> {
         let mut balance = self.balance;
         if self.deposit.batch_id < current_batch_id {
             balance = balance
                 .checked_add(self.deposit.amount)
                 .ok_or_else(|| anyhow!("math overflow"))?;
         }
+        if self.proceeds.batch_id < current_batch_id {
+            balance = balance
+                .checked_add(self.proceeds.increase)
+                .ok_or_else(|| anyhow!("math overflow"))?
+                .checked_sub(self.proceeds.decrease)
+                .ok_or_else(|| anyhow!("math underflow"))?;
+        }
         Ok(balance)
     }
 
-    fn apply_existing_deposit(&mut self, current_batch_id: BatchId) -> Result<()> {
+    fn apply_existing_deposit_and_proceeds(&mut self, current_batch_id: BatchId) -> Result<()> {
```

Comment:
Is there no analogue method which "applied existing withdraw"? Or should withdraws be included here?

Reply:
Withdraw Requests only get applied in the `withdraw` function. That function already existed so this diff doesn't show it.

Reply:
In that case, I am unclear on why deposits are handled together with proceeds, while withdraws are handled on their own.

Reply:
Deposits and proceeds can be applied whenever the batch id has increased because then they are final.
Withdraw requests on the other hand start being valid at some point at which we need to deduct them from the available balance (`get_balance`) but they only disappear when we see a Withdraw event.

Reply:
Right, yes! Thanks very much for clearing that up for me. Do you think a description like this might be useful to have somewhere within the code comments?

Reply:
I feel like the code is self describing in that way because I would just repeat what is already happening. But maybe I have worked with that code for so long that I don't  notice non obvious cases.